### ISSUE TITLE
[nvmf_filesystem_no_incapsule]  keep alive timeout when disconnecting host 

### ISSUE THREAD

**Issue by wanghailiangx** [spdk/spdk#1900]
************************************
START TEST nvmf_filesystem_no_incapsule
************************************

## Link to the failed CI build

https://ci.spdk.io/results/autotest-per-patch/builds/41867/archive/nvmf-tcp-phy-autotest/build.log

## Execution failed at

 13:46:01	# nvme connect -t tcp -n nqn.2016-06.io.spdk:cnode1 -a 10.0.0.2 -s 4420
 13:46:02	# waitforserial SPDK00000000000001
 13:46:02	# local i=0
 13:46:02	# local nvme_device_counter=1
 13:46:02	# [[ -n '' ]]
  13:46:02	# lsblk -l -o NAME,SERIAL
  13:46:02	# grep -c SPDK00000000000001
 13:46:02	# '[' 0 -lt 1 ']'
 13:46:02	# '[' 0 -lt 15 ']'
 13:46:02	# i=1
 13:46:02	# echo 'Waiting for devices'
Waiting for devices
 13:46:02	# sleep 1
  13:46:03	# lsblk -l -o NAME,SERIAL
  13:46:03	# grep -c SPDK00000000000001
 13:46:03	# '[' 1 -lt 1 ']'
  13:46:03	# lsblk -l -o NAME,SERIAL
  13:46:03	# grep -c SPDK00000000000001
 13:46:03	# [[ 1 -lt 1 ]]
 13:46:03	# return 0
  13:46:03	# lsblk -l -o NAME,SERIAL
  13:46:03	# grep -oP '([\w]*)(?=\s+SPDK00000000000001)'
 13:46:03	# nvme_name=nvme0n1
 13:46:03	# mkdir -p /mnt/device
 13:46:03	# parted -s /dev/nvme0n1 mklabel msdos mkpart primary 0% 100%
Warning: The resulting partition is not properly aligned for best performance: 1s % 2048s != 0s
 13:46:03	# partprobe
**[2021-04-14 13:46:21.742916] ctrlr.c: 185:nvmf_ctrlr_keep_alive_poll: *NOTICE*: Disconnecting host nqn.2014-08.org.nvmexpress:uuid:6e5bc17eae3c4b4cb5b6c885235fa21d from subsystem nqn.2016-06.io.spdk:cnode1 due to keep alive timeout.
Build timed out (after 15 minutes). Marking the build as failed.
Build was aborted**
[PostBuildScript] - [INFO] Executing post build scripts.
[nvmf-tcp-phy-autotest] $ /usr/bin/env bash /tmp/jenkins4330193478955321076.sh
Start stage post_get_build_log.sh

**Comment by changpe1** [comment]
Patch for the failure is here : https://review.spdk.io/gerrit/c/spdk/spdk/+/6932

**Comment by changpe1** [comment]
[Bug Scrub] @wanghailiangx Can you confirm that your patch can work with E810 card?

**Comment by wanghailiangx** [comment]
@changpe1 Yes, we can see tcp packets on one of interface when run fio with E810 card. If we remove the patch, we can’t see tcp packets on the interface.
```
# tcpdump -nnt -i ens786f1
IP 10.0.0.2.4420 > 10.0.0.1.50470: Flags [P.], seq 1494402913:1494406577, ack 25876441, win 52883, options [nop,nop,TS val 2779502887 ecr 2300876011], length 3664
IP 10.0.0.2.4420 > 10.0.0.1.50470: Flags [P.], seq 1494406577:1494423033, ack 25876441, win 52883, options [nop,nop,TS val 2779502887 ecr 2300876011], length 16456
IP 10.0.0.1.50470 > 10.0.0.2.4420: Flags [.], ack 1494406577, win 24479, options [nop,nop,TS val 2300876011 ecr 2779502887], length 0
IP 10.0.0.1.50470 > 10.0.0.2.4420: Flags [.], ack 1494423033, win 24567, options [nop,nop,TS val 2300876011 ecr 2779502887], length 0
IP 10.0.0.1.50470 > 10.0.0.2.4420: Flags [P.], seq 25877089:25878097, ack 1494423033, win 24567, options [nop,nop,TS val 2300876011 ecr 2779502887], length 1008
IP 10.0.0.2.4420 > 10.0.0.1.50470: Flags [.], seq 1494423033:1494453441, ack 25877089, win 52883, options [nop,nop,TS val 2779502887 ecr 2300876011], length 30408
```

**Comment by changpe1** [comment]
I didn't see such error without your patch, so please verify your patch in CI environment multiple times to see if this is intermittent failure.

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @wanghailiangx. log: https://ci.spdk.io/public_build/autotest-per-patch_42484.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @dong-liuliu. log: https://ci.spdk.io/public_build/autotest-per-patch_43051.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @tomzawadzki. log: https://ci.spdk.io/public_build/autotest-per-patch_43130.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @tomzawadzki. log: https://ci.spdk.io/public_build/autotest-per-patch_43128.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @jimharris. log: https://ci.spdk.io/public_build/autotest-per-patch_43181.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @jimharris. log: https://ci.spdk.io/public_build/autotest-per-patch_43195.html

**Comment by wanghailiangx** [comment]
Please see my comments in https://github.com/spdk/spdk/issues/1906. These two should be the same problem.

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @yidong0635. log: https://ci.spdk.io/public_build/autotest-per-patch_43407.html

**Comment by yidong0635** [comment]
Can we change  from the host side ?
that means to increase the keep alive timer, this will not cause CI testing time, Just enlarge the
tolerance of host connection . https://review.spdk.io/gerrit/c/spdk/spdk/+/7605. Apply this on other failing patch caused by this issue. It works .

**Comment by jimharris** [comment]
Hi @yidong0635,

I have a couple of concerns with just increasing the keep alive timer:

1) it just works around the problem - I think we should try to understand more *why* the keep alive timer is expiring before we just increase it
2) this issue is intermittent, so re-running a patch once and seeing it pass does not necessarily indicate it will always pass

-Jim

**Comment by wanghailiangx** [comment]
@yidong0635 When I repeated this issue on local occasionally, I change "keep_alive_timeout_ms" according to your patch, but it has no effect, in the process of running FIO, IP will still be lost.

**Comment by yidong0635** [comment]
@jimharris @yidong0635 yes, the root cause of this issue is still need other one to investigate. 
I just observed this keep alive timer timeout issue a long time. And most of the questions, they can be covered by changed the 
keep alive timer from the host side. This wastes us a lot time.  The keep alive timer is too short in real test environment some times. before we have done some experiments about this , so expose this item for fine tune from the host side, we know that 
spec defines it's 1 second . But that easily causes intermittent failure.

**Comment by wanghailiangx** [comment]
Now I don't repeat this problem locally. The reason why I can repeat it before is that I run the test immediately after reloading the ice driver. The time is not long enough, resulting in the loss of the set IP.

**Comment by wanghailiangx** [comment]
IP lost in fio and keep alive timeout, this can be reproduced sometimes on condition that:
1.  just run fio such as ./test/nvmf/host/fio.sh --transport=tcp --iso
2. systemctl start NetworkManager.service
3. ifup ens786f0 (ens786f0 is tcp network card target interface)
4. Run the above fio again and there is an error keep alive timeout.

The solution as a workaround is:
Before running nvme TCP test, excute systemctl stop networkmanager.service
When TCP test run done, start systemctl networkmanager.service. This is effective on local test.
ip link set ens786f0 down/up, ifdown/ifup have no effect.

If there is a better way, please comment in https://review.spdk.io/gerrit/c/spdk/spdk/+/7573

**Comment by wanghailiangx** [comment]
According to the tips Ziye gave, I verified the following methods, configuring a device as unmanaged in NetworkManager. It is effective for this issue. 
https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/configuring_and_managing_networking/configuring-networkmanager-to-ignore-certain-devices_configuring-and-managing-networking#permanently-configuring-a-device-as-unmanaged-in-networkmanager_configuring-networkmanager-to-ignore-certain-devices
```
# nmcli device status
DEVICE  TYPE      STATE      CONNECTION
enp1s0  ethernet  unmanaged  --
```

**Comment by optimistyzy** [comment]
So I think that if there is NIC cards equipped for test (Using the IPs allocated by our script), then we should write some scripts to remove those cards from network service manager.  Then this solution will not affect the DHCP IP configuration and retrieval  of the embedded NIC for management purpose. It is much better to just stop the network service manager service during the running the tests.

**Comment by changpe1** [comment]
[Bug Scrub] The issue was due to the network issue, but we need to figure out which function is blocked during the tests, because perf only test with 10 seconds, while the keep alive timeout is 20 seconds.

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @changpe1. log: https://ci.spdk.io/public_build/autotest-per-patch_43765.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @changpe1. log: https://ci.spdk.io/public_build/autotest-per-patch_43759.html

**Comment by spdkci** [comment]
Another instance of this failure. Reported by @changpe1. log: https://ci.spdk.io/public_build/autotest-per-patch_43763.html

**Comment by optimistyzy** [comment]
> [Bug Scrub] The issue was due to the network issue, but we need to figure out which function is blocked during the tests, because perf only test with 10 seconds, while the keep alive timeout is 20 seconds.

It could be divided into two issues. The first issue is the IP lost issue, and the second is the hang issue. HaiLiang's work is to address the IP lost issue.

**Comment by changpe1** [comment]
since the E810 NIC card has been removed from Network Management, let's keep this issue open for a while to see if it still exist.

**Comment by changpe1** [comment]
closing this issue for now as I don't see same failure recently.
